bftengine/src/bftengine/TimeServiceManager.hpp

```diff
@@ -0,0 +1,98 @@
+// Concord
+//
+// Copyright (c) 2018-2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").  You may not use this product except in
+// compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include "Replica.hpp"
+#include "ReplicaConfig.hpp"
+#include "TimeServiceResPageClient.hpp"
+#include "assertUtils.hpp"
+#include "messages/ClientRequestMsg.hpp"
+#include "serialize.hpp"
+#include <chrono>
+#include <cstdlib>
+#include <limits>
+#include <memory>
+
+namespace bftEngine {
+
+template <typename ClockT = std::chrono::system_clock>
+class TimeServiceManager {
+ public:
+  explicit TimeServiceManager(const ReplicaConfig& config) : config_(config) {}
+  ~TimeServiceManager() = default;
+  TimeServiceManager(const TimeServiceManager&) = delete;
+
+  // Checks if the new time is less or equal to the one reserved pages,
+  // if this is the case, returns reserved pages time + epsilon
+  // otherwise, returns the new time
+  [[nodiscard]] ConsensusTime compareAndSwap(ConsensusTime new_time) {
+    auto last_timestamp = client_.getLastTimestamp();
+    if (new_time > last_timestamp) {
+      client_.setLastTimestamp(new_time);
+      return new_time;
+    }
+
+    LOG_INFO(TS_MNGR,
```

Comment:
Should be at DEBUG level IMHO.

Reply:
Discussed offline: the situation is not supposed to happen very frequently, if it does happen then action should be taken.